### PR TITLE
Replace fonts build system

### LINKED ISSUE #13: Reduce the number of available ligatures to 1 per icon
Currently the font supports three ligatures (for most) icons, limiting this to just one can reduce the size of the output font. It's not much, as the table below shows, but it only becomes more beneficial over time. Also, I don't really see a reason to support three input values for the same icon, especially with #6 in mind.

| File | Size before (in bytes) | Size after (in bytes) |
|---|---|---|
| `SimpleIcons.woff` | `222,084` | `216,236` |
| `SimpleIcons.woff2` | `192,436` | `188,184` |

### LINKED ISSUE #16: Find a approriate solution for icons too complex for fonts
As per https://github.com/simple-icons/simple-icons-font/pull/12#discussion_r305582647 we should find a solution for icons that are too large/complex (e.g. the [Elsevier](https://github.com/simple-icons/simple-icons/blob/1978e05c5ddd115ef445099cbe48bffa72e0f1e6/icons/elsevier.svg#L1) icon).

Whether that is somehow getting it in the font (e.g. by reducing the complexity) or through documentation is open for discussion.

Relatedly we should look into proper error messaging when building the font and, if we stick with removing icons that are too large from the font, find a proper way to configure that, probably in an external file(?)

### LINKED ISSUE #30: Overlapping names confuses ligatures and wrong icons are displayed
Have a look at the image below, you might see the logo for the C and R programming languages quite often 😕 

This is because the ligature for e.g. "CoronaRenderer" (highligted in the image) contains a C and 4 R's and instead of showing the ligature for it shows 1 for the C programming language and 4 for the R programming language.

I'm not sure if fonts can be somehow tweaked to look for the longest ligatures first... One interesting observation is that the different browsers show different icons differently. For example the Corona Renderer logo doesn't show in Firefox (below) but works in Chrome.

image

### LINKED ISSUE #41: Increase functionality of test page
Would be interesting to have a toggle, besides light/dark background, on the test page that switches between:

- [ ] The class variant and the ligature variant
- [x] Black icons and brand colored icons - #43

### PR DESCRIPTION
**Pull request by mondeja**
This pull creates a font build system at lower level, replacing the one that was using `icon-font-buildr`.

- Removes the need of SVGO usage before build.
- Closes #30 because now the icons ligatures are built using HEX unicode characters.
- Closes #16 because even the Elsevier icon is rendering correctly.
- Closes #41.
- Contributes to #14.

### CODE REVIEW

File: preview/html/testpage.pug [hunk 1]
```diff
@@ -3,7 +3,7 @@ html
 	head
 		meta(charset='utf-8')
 
-		link(href='font/font.css', rel='stylesheet', type='text/css')
+		link(href='font/SimpleIcons.css', rel='stylesheet', type='text/css')
```
Comment (ericcornelissen) [inline preview/html/testpage.pug:6]: I would prefer the filename to be all lowercase (either _"simpleicons.css"_ or _"simple-icons.css"_) Could you update that?

File: scripts/build.js [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+
+const buildSimpleIconsSvgFontFile = () => {
+  let fileContent = '<?xml version="1.0" standalone="no"?>\n'
+                  + '<!DOCTYPE svg PUBLIC "-//W3C//DTD SVG 1.1//EN" "http://www.w3.org/Graphics/SVG/1.1/DTD/svg11.dtd">\n'
+                  + '<svg xmlns="http://www.w3.org/2000/svg">\n'
+                  + '<metadata>CC0 1.0 Universal | simple-icons contributors</metadata>\n'
+                  + '<defs>\n'
+                  + '<font id="Simple Icons" horiz-adv-x="1200">\n'
+                  + '<font-face font-family="Simple Icons" units-per-em="1200" ascent="-1" descent="1200" />\n'
+                  + '<missing-glyph horiz-adv-x="0" />\n';
```
Comment (ericcornelissen) [inline scripts/build.js:22]: How about :arrow_down: ?

```suggestion
  let fileContent = `<?xml version="1.0" standalone="no"?>
                  <!DOCTYPE svg PUBLIC "-//W3C//DTD SVG 1.1//EN" "http://www.w3.org/Graphics/SVG/1.1/DTD/svg11.dtd">
                  <svg xmlns="http://www.w3.org/2000/svg">
                  <metadata>CC0 1.0 Universal | simple-icons contributors</metadata>
                  <defs>
                  <font id="Simple Icons" horiz-adv-x="1200">
                  <font-face font-family="Simple Icons" units-per-em="1200" ascent="-1" descent="1200"/>
                  <missing-glyph horiz-adv-x="0"/>`;
```

File: scripts/build.js [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+      String.fromCodePoint(startUnicode++)).map(
+        point => '&#x' + point.toString(16).toUpperCase() + ';')[0];
+    if (usedUnicodes.includes(unicode)) {
+      throw Error(`Unicodes must be unique. Found '${unicode}' repeated`)
+    }
+    
+    let icon = SimpleIcons[iconTitle];
+    const verticalTransformedPath = SVGPath(icon.path).scale(50, -50).toString();
+    
+    fileContent += `<glyph glyph-name="${icon.slug}" unicode="${unicode}" d="${verticalTransformedPath}" horiz-adv-x="1200" />\n`;
```
Comment (ericcornelissen) [inline scripts/build.js:40]: Are the newlines necessary? Not sure if it would reduce the font-size if we remove them but that could be worth considering.

```suggestion
    fileContent += `<glyph glyph-name="${icon.slug}" unicode="${unicode}" d="${verticalTransformedPath}" horiz-adv-x="1200" />`;
```

File: scripts/build.js [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+  
+  let startUnicode = 0xea01;
+  let usedUnicodes = [];
+  
+  let unicodeHexBySlug = {}
+  
+  for (let iconTitle in SimpleIcons) {
+    let unicode = ucs2.decode(
+      String.fromCodePoint(startUnicode++)).map(
+        point => '&#x' + point.toString(16).toUpperCase() + ';')[0];
```
Comment (ericcornelissen) [inline scripts/build.js:32]: This part is a bit hard to follow, and I would like to better understand what's going on. Are the suggestions below correct? If so, could you update this and the surrounding code?

```suggestion
    let nextUnicode = ucs2.decode(String.fromCodePoint(unicodeIndex++))[0];
    let unicodeString = nextUnicode.map(point => `&#x${point.toString(16).toUpperCase()};`);
```

File: scripts/build.js [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+      path.resolve(__dirname, '..', 'preview', 'css', 'base.css'));
+  
+  for (let slug in unicodeHexBySlug) {
+    let icon = unicodeHexBySlug[slug];
+    
+    // &#xF26E; -> \f26e
+    const cssDecodedUnicode = icon.unicode.replace('&#x', '\\').replace(';', '').toLowerCase();
+    simpleIconsCss += '\n'
+                    +   `.simpleicons-${slug}::before { content: "${cssDecodedUnicode}"; }\n`
+                    +   `.simpleicons-${slug}.simpleicons--color::before { color: #${icon.hex}; }\n`;
```
Comment (ericcornelissen) [inline scripts/build.js:68]:
```suggestion
    simpleIconsCss += `
                    .simpleicons-${slug}::before { content: "${cssDecodedUnicode}"; }
                    .simpleicons-${slug}.simpleicons--color::before { color: #${icon.hex}; }`;
```

File: scripts/build.js [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+
+const buildSimpleIconsCssFile = (unicodeHexBySlug) => {
+  let simpleIconsCss = fs.readFileSync(
+      path.resolve(__dirname, '..', 'preview', 'css', 'base.css'));
+  
+  for (let slug in unicodeHexBySlug) {
+    let icon = unicodeHexBySlug[slug];
+    
+    // &#xF26E; -> \f26e
+    const cssDecodedUnicode = icon.unicode.replace('&#x', '\\').replace(';', '').toLowerCase();
```
Comment (ericcornelissen) [inline scripts/build.js:65]: Maybe just define a helper function at the top of this file that does this called something like `cssDecodeUnicode()`?
```suggestion
    const cssDecodedUnicode = cssDecodeUnicode(icon.unicode);
```

File: scripts/build.js [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+                    +   `.simpleicons-${slug}.simpleicons--color::before { color: #${icon.hex}; }\n`;
+  }
+  
+  const outputFilePath = path.join(DISTDIR, 'SimpleIcons.css');
+  fs.writeFileSync(outputFilePath, simpleIconsCss);
+  console.log(`'${outputFilePath}' file built`);
+}
+
+const buildSimpleIconsTtfFontFile = () => {
+  const ttf = svg2ttf(fs.readFileSync(path.join(DISTDIR, 'SimpleIcons.svg'), 'utf8'), {});
```
Comment (ericcornelissen) [inline scripts/build.js:77]: Would prefer not having to read this after it has been created and instead just pass the `SimpleIcons.svg` content as an argument.

File: scripts/build.js [hunk 8]
```diff
@@ -0,0 +1,110 @@
...
+  const outputFilePath = path.join(DISTDIR, 'SimpleIcons.ttf');
+  fs.writeFileSync(outputFilePath, new Buffer(ttf.buffer));
+  console.log(`'${outputFilePath}' file built`);
+}
+
+const buildSimpleIconsWoffFontFile = () => {
+  var ttf = new Uint8Array(fs.readFileSync(path.join(DISTDIR, 'SimpleIcons.ttf')));
```
Comment (ericcornelissen) [inline scripts/build.js:84]: Would prefer not having to read this after it has been created and instead just pass the `SimpleIcons.ttf` content as an argument.

File: scripts/build.js [hunk 9]
```diff
@@ -0,0 +1,110 @@
...
+  var woff = new Buffer(ttf2woff(ttf).buffer);
+  const outputFilePath = path.join(DISTDIR, 'SimpleIcons.woff');
+  fs.writeFileSync(outputFilePath, woff);
+  console.log(`'${outputFilePath}' file built`);
+}
+
+const buildSimpleIconsWoff2FontFile = () => {
+  const input = fs.readFileSync(path.join(DISTDIR, 'SimpleIcons.ttf'));
```
Comment (ericcornelissen) [inline scripts/build.js:92]: Would prefer not having to read this after it has been created and instead just pass the `SimpleIcons.ttf` content as an argument.

**Review by ericcornelissen — CHANGES_REQUESTED** [review]
Looks promising :+1: Got a few comments below and one more comment that doesn't fit inline, can you update [buildTestpage.js](https://github.com/simple-icons/simple-icons-font/blob/06815453e4aa3edbd2278257d04878a15cf7006b/scripts/buildTestpage.js#L10-L11) so that _Elsevier_ is included in the testpage?

File: package.json [hunk 10]
```diff
@@ -32,10 +30,12 @@
   },
   "devDependencies": {
     "anywhere": "1.5.0",
-    "icon-font-buildr": "1.3.4",
-    "js-yaml": "3.14.0",
     "pug": "3.0.0",
-    "simple-icons": "2.3.0",
-    "svgo": "1.3.2"
+    "punycode": "2.1.1",
+    "simple-icons": "4.1.0",
```
[the remark this answers is not in the document]
Comment (mondeja) [reply package.json:35]: I thought that a release wouldn't be published until the whole font build system was in order and that the next release would be in accordance with simple-icons version, but yes, maybe it's more comfortable go step by step from old ones to the new version :+1:

File: scripts/build.js (hunk 3 above)
Comment (mondeja) [reply scripts/build.js:40]: I assumed that we would minify at launch, but there is not much to see inside so I could build it already minified, but also the `<metadata>`, `<defs>`, `<font>`... part?

File: scripts/build.js (hunk 4 above)
Comment (mondeja) [reply scripts/build.js]: Yes, are correct, but I've improved it a bit:

```javascript
let nextUnicode = ucs2.decode(String.fromCodePoint(startUnicode++));
let unicodeString = nextUnicode.map(point => `&#x${point.toString(16).toUpperCase()};`).join('');
```
- I've used `startUnicode` instead of `unicodeIndex` because is an Unicode character in HEX format, not an index.
- This code is extracted from `svgicons2svgfont` (dependency of `icon-font-buildr`), from [here](https://github.com/nfroidure/svgicons2svgfont/blob/b8ed724ec20793edfeec913f7184fdbab16909f2/src/index.js#L488-L491), [here](https://github.com/nfroidure/svgicons2svgfont/blob/94968febca43b10e6b498a111198d70e1790c1ca/src/metadata.js#L61) and [here](https://github.com/nfroidure/svgicons2svgfont/blob/94968febca43b10e6b498a111198d70e1790c1ca/src/metadata.js#L15-L16). I'm not 100% sure it's correct, but the theory is to iterate through the unicode characters one by one and insert them encoded in the font file, one for each icon, then in the CSS insert them encoded with the format that CSS needs.

File: scripts/build.js (hunk 6 above)
Comment (mondeja) [reply scripts/build.js]: Sure :smile:

I've included in the CSS append, should keep it out for some reason?

```javascript
simpleIconsCss += `
.simpleicons-${slug}::before { content: "${cssDecodeUnicode(icon.unicode)}"; }
.simpleicons-${slug}.simpleicons--color::before { color: #${icon.hex}; }`;
```

File: scripts/build.js (hunk 7 above)
Comment (mondeja) [reply scripts/build.js]: Sorry for the mess, I've defined now built assets filepaths as global variales and passed the contents of the files as you have requested. I hope that this could be more readable.

File: scripts/build.js (hunk 6 above)
Comment (ericcornelissen) [reply scripts/build.js]: No that is fine :+1:

File: scripts/build.js (hunk 7 above)
Comment (ericcornelissen) [reply scripts/build.js]: No problem. It was quite readable but I do think it is much nicer now. The main reason I requested it is because reading it from disk is a bit of a waste of time if you already have it in memory anyway.

File: package.json (hunk 10 above)
[the remark this answers is not in the document]
Comment (ericcornelissen) [reply package.json]: I agree with your first statement, I would just try to make sure there is no accidental release at v4 to early.

I don't think it is necessary to retro-actively create a release for every simple-icons version, but I do think it could be worthwhile to create at least one for the last v2 and v3 releases.

File: scripts/build.js (hunk 4 above)
Comment (ericcornelissen) [reply scripts/build.js]: > I've used startUnicode instead of unicodeIndex because is an Unicode character in HEX format, not an index.

I suggested changing the name since it used as a running "index"-like value, not just the starting value. Would still like to see that changed but indeed, index doesn't seem like the best word...


> This code is extracted from svgicons2svgfont (dependency of icon-font-buildr), from here, here and here. I'm not 100% sure it's correct, but the theory is to iterate through the unicode characters one by one and insert them encoded in the font file, one for each icon, then in the CSS insert them encoded with the format that CSS needs.

I figured something like that was going on. I think it is quite clear now what it does (i.e. iterating over unicode characters), I just think the original 3-line spanning statement was a bit much to parse.

File: scripts/build.js (hunk 3 above)
Comment (ericcornelissen) [reply scripts/build.js]: > I assumed that we would minify at launch

Fair enough, I would agree. But, actually, do we need to include `SimpleIcons.svg` when publishing to NPM?


> but also the `<metadata>`, `<defs>`, `<font>` ... part?

I would say not, but I'm not sure. The main reason I asked it here is because it serves no purpose in the code. With the other part at least it makes it a bit more readable than a single long line...

We could consider doing [something similar to the main repo](https://github.com/simple-icons/simple-icons/blob/develop/scripts/templates/icon-object.js) where we have a template file for the `SimpleIcons.svg` file and just insert the long string with all `<glyph>`s at the end. What do you think about that?

Comment (mondeja) [reply scripts/build.js]: > But, actually, do we need to include SimpleIcons.svg when publishing to NPM?

FontAwesome does that: `.svg`, `.eot`, `.ttf`, `.woff` and `.woff2`. I suppose that could be useful to someone.

> We could consider doing something similar to the main repo where we have a template file for the SimpleIcons.svg file and just insert the long string with all <glyph>s at the end. What do you think about that?

Yes, good idea. I will do that removing all newlines after glyphs, keeping the metadata readable :+1:

Comment (mondeja) [reply scripts/build.js]: Done :wrench: :sparkles:

Comment (ericcornelissen) [reply scripts/build.js]: > FontAwesome does that: .svg, .eot, .ttf, .woff and .woff2. I suppose that could be useful to someone.

Hmm, still not 100% convinced but this doesn't need to block this PR from being merged.

I did some local testing and it seems each font works fine without the `.svg` there. Given that the SVGs are already available through the main package I feel like it is a bit redundant to include the SVG file here as well :thinking: It's not much (1.6MB) but I think we should consider adding the following to `.npmignore`:

```
/font/SimpleIcons.svg
```

If not here, I will open an issue for it so that we can decide on it later.

File: scripts/build.js [hunk 11]
```diff
@@ -0,0 +1,112 @@
...
+.simpleicons-${slug}.simpleicons--color::before { color: #${icon.hex}; }`;
+  }
+
+  fs.writeFileSync(CSS_OUTPUT_FILEPATH, simpleIconsCss);
+  console.log(`'${CSS_OUTPUT_FILEPATH}' file built`);
+}
+
+const buildSimpleIconsTtfFontFile = (svgFileContent) => {
+  const ttf = svg2ttf(svgFileContent, {});
+  const ttfFileContent = new Buffer(ttf.buffer);
```
Comment (ericcornelissen) [inline scripts/build.js:82]: Just noticed a deprecation warning when building:

```
(node:16498) [DEP0005] DeprecationWarning: Buffer() is deprecated due to security and usability issues. Please use the Buffer.alloc(), Buffer.allocUnsafe(), or Buffer.from() methods instead
```

Can you update this and other instances of Buffer in this file? Seems to work just fine with `Buffer.from`.

```suggestion
  const ttfFileContent = new Buffer.from(ttf.buffer);
```

File: package.json [hunk 12]
```diff
@@ -32,10 +30,12 @@
...
-    "icon-font-buildr": "1.3.4",
-    "js-yaml": "3.14.0",
     "pug": "3.0.0",
+    "punycode": "2.1.1",
     "simple-icons": "2.3.0",
-    "svgo": "1.3.2"
+    "svg2ttf": "5.0.0",
+    "svgpath": "2.3.0",
+    "ttf2woff": "2.0.2",
+    "ttf2woff2": "3.0.0"
```
Comment (ericcornelissen) [inline package.json:39]: Minor request, could you include a commit where you run `npm audit fix`/`npm update lodash --depth 6`?

File: scripts/build.js (hunk 3 above)
Comment (mondeja) [reply scripts/build.js]: OK, I will do and, if someones need, could be discussed in aonther issue.

Comment (mondeja) [reply scripts/build.js]: Done! :wrench:

File: package.json (hunk 12 above)
Comment (mondeja) [reply package.json:39]: Sure, it's done.

**Review by ericcornelissen — APPROVED** [review]

**Comment by ericcornelissen** [comment]
Amazing work @mondeja :tada: Should get us closer to getting this project up and running :+1:
